### ISSUE TITLE
Support for carret and tilde requirements

### ISSUE THREAD

**Issue by seppzer0** [astral-sh/uv#6590]
Hi! I'm migrating from Poetry, and I'm very much excited for the functionality that is currently in uv 🌟 

My current interest is in dependency version specification. In my understanding, carret and tilde requirements belong exclusively to Poetry's [dependency specification](https://python-poetry.org/docs/dependency-specification/), meaning that because of this non-standard format it would make it impossible to use them in the `[project]` section of pyproject.toml file.

To be clear about this, both carret and tilde requirements are essentially aliases to standard version range formats (e.g., `^1.2` is the same as `>=1.2.0 <2.0.0`), which means that technically there is nothing blocking me from using standard format in the `[project]` section to achieve desired result. However, having such short aliases is indeed quite convenient and does add a tidier look in a big list of dependencies. The priority of this functionality would be low, but also a nice touch.

My suggestion is to add both carret and tilde requirements support for both dev and regular dependencies in `[tool.uv]` section of pyproject.toml . This would not break the standard pattern usage in `[project]`, but also add this type of functionality in general if one chooses to use it.

What do you think? 📝

**Comment by hauntsaninja** [comment]
Note PEP 440 lets you use `dependency==1.2.*` or `dependency~=1.2`.

(Personal opinions: I don't really see the need for additional dep specification forms beyond those ones. I'm really glad that uv chose not to default to upper bounding dependencies)

**Comment by seppzer0** [comment]
@hauntsaninja ,

> Note PEP 440 lets you use dependency==1.2.* or dependency~=1.2.

I did not know that. Thank you, this is something I can switch to!

> glad that uv chose not to default to upper bounding dependencies

Yes, I think this is a problem Poetry used to have (or maybe still has, not sure) and agree that it's nice uv doesn't have that. To be fair though, upper bounding reduces the risk of your application breaking with time, which for my experience I assume by default and why I got interested in carret requirements in uv. Opt-in, that is.

To summarize, this issue can be closed. I was unaware PEP 440 already has the essential idea of Poetry's carret + tilde requirements, and I hope I did not take too much of everyone's time :)
